apps/lib/chain_pusher/solana.go

```diff
@@ -0,0 +1,524 @@
+		return nil, err
+	}
+
+	contractPubKey, err := solana.PublicKeyFromBase58(contractAddr)
+	if err != nil {
+		logger.Fatal().Err(err).Msg("Invalid contract address")
+		return nil, err
+	}
+
+	payer, err := solana.PrivateKeyFromSolanaKeygenFile(privateKeyFile)
+	if err != nil {
+		logger.Fatal().Err(err).Msg("Failed to parse private key")
+		return nil, err
+	}
+
+	assetConfig, err := LoadConfig(assetConfigFile)
+	if err != nil {
+		logger.Fatal().Err(err).Msg("Failed to load asset config")
+		return nil, err
+	}
+
+	feedAccounts := make(map[InternalEncodedAssetId]solana.PublicKey)
+	for _, asset := range assetConfig.Assets {
+		encodedAssetIdBytes, err := hexStringToByteArray(string(asset.EncodedAssetId))
+		if err != nil {
+			logger.Fatal().Err(err).Str("assetId", fmt.Sprintf("%v", asset.AssetId)).Msg("Failed to convert encoded asset ID to bytes")
+			return nil, err
+		}
+		//derive pda
+		feedAccount, _, err := solana.FindProgramAddress(
+			[][]byte{
+				[]byte("stork_feed"),
+				encodedAssetIdBytes,
+			},
+			contractPubKey,
+		)
+		if err != nil {
+			logger.Fatal().Err(err).Str("assetId", fmt.Sprintf("%v", asset.AssetId)).Msg("Failed to derive PDA for feed account")
+			return nil, err
+		}
+
+		encodedAssetId := InternalEncodedAssetId(encodedAssetIdBytes)
+		feedAccounts[encodedAssetId] = feedAccount
+	}
+
+	treasuryAccounts := make(map[uint8]solana.PublicKey)
+	for i := 0; i < 256; i++ {
+		treasuryAccount, _, err := solana.FindProgramAddress(
+			[][]byte{[]byte("stork_treasury"), {uint8(i)}}, contractPubKey)
+		if err != nil {
+			logger.Fatal().Err(err).Uint8("treasuryId", uint8(i)).Msg("Failed to derive PDA for treasury account")
+			return nil, err
+		}
+		treasuryAccounts[uint8(i)] = treasuryAccount
+	}
+
+	configAccount, _, err := solana.FindProgramAddress(
+		[][]byte{[]byte("stork_config")}, contractPubKey)
+	if err != nil {
+		logger.Fatal().Err(err).Msg("Failed to derive PDA for config account")
+		return nil, err
+	}
+
+	confirmationInChan := make(chan solana.Signature)
+	confirmationOutChan := make(chan solana.Signature)
+
+	contract.SetProgramID(contractPubKey)
+	sci := &SolanaContractInteracter{
+		logger:              logger,
+		client:              client,
+		wsClient:            wsClient,
+		contractAddr:        contractPubKey,
+		feedAccounts:        feedAccounts,
+		treasuryAccounts:    treasuryAccounts,
+		configAccount:       configAccount,
+		payer:               payer,
+		limiter:             limiter,
+		pollingFrequencySec: pollingFreqSec,
+		batchSize:           batchSize,
+		confirmationInChan:  confirmationInChan,
+	}
+
+	avgBlockTime, err := sci.getRecentAvgBlockTime()
+	if err != nil {
+		logger.Fatal().Err(err).Msg("Failed to get recent average block time")
+		return nil, err
+	}
+	numWorkers := int(math.Ceil(math.Ceil(float64(len(assetConfig.Assets))/float64(sci.batchSize)) * avgBlockTime.Seconds()))
+	//add 1 for safety
+	numWorkers += 1
```

Comment:
Can you walk me through the reasoning behind this formula? If these workers are just logging whether a transaction was confirmed or not, couldn't we just have a single worker working through the queue of signatures at whatever speed? or at least a fixed size pool. It makes me a little nervous that we could spin up a huge number of threads here.

Reply:
see [here](https://github.com/Stork-Oracle/stork-external/pull/37#discussion_r1826026127): 

the math on what is probably a bad case scenario, lets say 40 assets, 500ms blocktimes, 1 second batching window, is still only 5 workers. A more realistic scenario, 10 assets, 400ms blocktimes, 5 second batching window, is just 1 worker. Assuming the prices of all assets are always being pushed (which depends on volatility and is a worst case), this calculates the minimum number of workers required for them to always be able to keep up with queue. The unbounded queue is still maintained as a buffer against impermanent slowdowns we cant control like blocktime slowdowns and slow rpc responses. It is possible that the rpc calls would be fast enough for one worker to churn through the queue sequentially and still keep up, however from my understanding go routines are so lightweight that its [practical for 10s to 100s of thousands to be running without major performance hit](https://go.dev/doc/faq#goroutines), so I figure setting the number of workers at sort of the upper bound for what is necessary to prevent blocking new updates or queue growth creep is fine, especially if its pretty much on the order of 10 or less in practical scenarios.